File: Data Structures & Algorithms/lowest-common-ancestor-in-binary-search-tree/submission-2.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def lowestCommonAncestor(self, root: TreeNode, p: TreeNode, q: TreeNode) -> TreeNode:
        # Start at top
        cur = root

        # Note that this tree is kinda sorted. Left of a node is smaller,
        # right of a node is larger than the current value.
        while cur:
            # if both nodes are larger than the current, it might be a 
            # right subnode
            if p.val > cur.val and q.val > cur.val:
                cur = cur.right
            # if both nodes smaller than currnet, must be left subnode
            elif p.val < cur.val and q.val < cur.val:
                cur = cur.left
            # if it's a split, or indeed the actual ancestor, just return cur
            else:
                return cur

File: Data Structures & Algorithms/level-order-traversal-of-binary-tree/submission-2.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    vector<vector<int>> levelOrder(TreeNode* root) {
        vector<vector<int>> levels;
        if (!root) return levels;
        
        queue<TreeNode*> q;
        q.push(root);

        while (!q.empty()) {
            vector<int> currLevel;
            int qSize = q.size();

            for (int i = qSize; i > 0; i--) {
                TreeNode* node = q.front();
                q.pop();

                if (node) {
                    currLevel.push_back(node->val);
                    q.push(node->left);
                    q.push(node->right);
                }
            }
            if (!currLevel.empty()) {
                levels.push_back(currLevel);
            }
        }

        return levels;
    }
};
